rtgui/threadutils.cc

```diff
@@ -136,12 +137,12 @@ void MyWriterLock::release ()
         return;
     }
 
-    Glib::Threads::Mutex::Lock lock (mutex.mutex);
+    std::lock_guard<std::mutex> lock (mutex.mutex);
 
     // Decrement the writer number first...
     if (--mutex.writerCount == 0) {
         // ...and if the writer count is zero again, we can wake up the next writer or reader.
-        mutex.cond.broadcast ();
+        mutex.cond.notify_one();  // notify_all ?
```

Comment:
Hm, we want multiple readers, so we should notify them all, don't we?

I asked my workmate to review this piece of code, too, and he noticed a possible imbalance in `MyReaderLock::acquire()`:

```c++
void MyReaderLock::acquire ()
{
    if (locked) {
        return;
    }

    std::lock_guard<std::mutex> lock (mutex.mutex);
// Precondition: A writer is writing, so writerCount != 0.
    if (mutex.writerCount == 0) {
        // There's no writer operating, we can increment the writer count which will lock writers.
        ++mutex.writerCount;
    } else if (mutex.readerCount == 0) {
// There is no active reader yet, so readerCount == 0.
        // The writer count is non null, but a reader can be the owner of the writer lock,
        // which will be the case if the reader count is not zero too.
// Now let two readers spin here.
        while (mutex.writerCount != 0) {
            mutex.cond.wait(mutex.mutex);
        }
// The writer leaves, both readers proceed.
        // Then, we can increment the writer count.
// The writerCount is incremented by 2. This leads to an imbalance in release().
        ++mutex.writerCount;
    }

    // Finally, we can increment the reader count as well.
    ++mutex.readerCount;

    locked = true;
}

void MyReaderLock::release ()
{
    if (!locked) {
        return;
    }

    std::lock_guard<std::mutex> lock (mutex.mutex);

    // decrement the writer number first...
    --mutex.readerCount;

    if (mutex.readerCount == 0) {
        // ...if no more reader, we decrement the writer count as well...
// Last reader leaves, but writerCount is only decremented by 1.
        --mutex.writerCount;

        // ...and signal the next waiting reader/writer that it's free
        mutex.cond.notify_one();  // notify_all ?
    }

    locked = false;
}
```

Is this plausible?

Reply:
`broadcast`  notifies all waiting threads, so `notify_all` is likely the proper answer.

Reply:
@Floessie 

About 
```
// Now let two readers spin here.
```

Isn't `std::lock_guard<std::mutex> lock (mutex.mutex);` protecting from this possibility ?